src/wallet/rpcwallet.cpp

```diff
@@ -305,6 +305,96 @@ UniValue sendtoaddress(const UniValue& params, bool fHelp)
     return wtx.GetHash().GetHex();
 }
 
+UniValue listaddresses(const UniValue& params, bool fHelp)
+{
+    if (!EnsureWalletIsAvailable(fHelp))
+        return NullUniValue;
+
+    if (fHelp)
+        throw runtime_error(
+            "listaddresses\n"
+            "\nLists the addresses tracked by this wallet.\n"
+            "\nResult:\n"
+            "[\n"
+            "  [\n"
+            "    {\n"
+            "      \"transparent\": { \"address\": \"t14oHp2v54vfmdgQ3v3SNuQga8JKHTNi2a1\", \"watchonly\": false },\n"
+            "    }\n"
+            "    {\n"
+            "      \"sprout\": { \"address\": \"ztbx5DLDxa5ZLFTchHhoPNkKs57QzSyib6UqXpEdy76T1aUdFxJt1w9318Z8DJ73XzbnWHKEZP9Yjg712N5kMmP4QzS9iC9\"},\n"
+            "    }\n"
+            "    {\n"
+            "      \"sapling\": { \"address\": \"ztbx5DLDxa5ZLFTchHhoPNkKs57QzSyib6UqXpEdy76T1aUdFxJt1w9318Z8DJ73XzbnWHKEZP9Yjg712N5kMmP4QzS9iC9\"},\n"
+            "    }\n"
+            "    ,...\n"
+            "  ]\n"
+            "  ,...\n"
+            "]\n"
+            "\nExamples:\n"
+            + HelpExampleCli("listaddresses", "")
+            + HelpExampleRpc("listaddresses", "")
+        );
+
+    LOCK2(cs_main, pwalletMain->cs_wallet);
+
+    KeyIO keyIO(Params());
+
+    UniValue ret(UniValue::VARR);
+
+    {
+        std::set<CTxDestination> transparentDests;
+        // Get the CTxDestination values for all the entries in the transparent address book
+        for (const std::pair<CTxDestination, CAddressBookData>& item : pwalletMain->mapAddressBook) {
+            transparentDests.insert(item.first);
+        }
+
+        // Ensure we have every address that holds a balance
+        for (const std::pair<CTxDestination, CAmount>& item : pwalletMain->GetAddressBalances()) {
+          transparentDests.insert(item.first);
```

Comment:
```suggestion
            transparentDests.insert(item.first);
```

Reply:
Under what circumstances could this loop add addresses that are not found by the preceding loop?